Fix -norpcwhitelist, -norpcallowip, and similar corner case behavior

ryanofsky opened the pull request:
The PR changes behavior of negated `-noseednode`, `-nobind`, `-nowhitebind`, `-norpcbind`, `-norpcallowip`, `-norpcwhitelist`, `-notest`, `-noasmap`, `-norpcwallet`, `-noonlynet`, and `-noexternalip` options, so negating these options just clears previously specified values doesn't have other side effects.

Negating options on the command line can be a useful way of resetting options that may have been set earlier in the command line or config file. But before this change, negating these options wouldn't fully reset them, and would have confusing and undocumented side effects (see commit descriptions for details). Now, negating these options just resets them and behaves the same as not specifying them.

Motivation for this PR is to fix confusing behaviors and also to remove incorrect usages of the `IsArgSet()` function. Using `IsArgSet()` tends to lead to negated option bugs in general, but it especially causes bugs when used with list settings returned by `GetArgs()`, because when these settings are negated, `IsArgSet()` will return true but `GetArgs()` will return an empty list. This PR eliminates all uses of `IsArgSet()` and `GetArgs()` together, and followup PR #17783 makes it an error to use `IsArgSet()` on list settings, since calling `IsArgSet()` is never actually necessary. Most of the changes here were originally made in #17783 and then moved here to be easier to review and avoid a dependency on #16545.

ryanofsky committed:
refactor: Avoid using IsArgSet() on -signetseednode and -signetchallenge list options

This commit does not change behavior because negation of -signetseednode and
-signetchallenge parameters has been disallowed since these were introduced in
#18267, so calling IsArgSet() is equivalent to checking if GetArgs() returns a
non-empty list.

ryanofsky committed:
Normalize inconsistent -noexternalip behavior

Treat specifying -noexternalip the same as not specifying -externalip, instead
of causing it to soft-set the -discover default to false.

Before this change, was -noexternalip basically an undocumented synonym for
-nodiscover.

After this change, specifying -noexternalip just clears previously specifed
-externalip options, restoring default behavior as if they were not were
specified.

The previous -noexternalip behavior wasn't neccessarily bad, but it was
undocumented, redundant with the -nodiscover option, and inconsistent with
behavior of other list options.

ryanofsky committed:
Fix nonsensical -nobind and -nowhitebind behavior

Treat specifying -nobind and -nowhitebind the same as not specifying -bind and
-whitebind values instead of causing them to soft-set -listen=1.

ryanofsky committed:
refactor: Avoid using IsArgSet() on -connect list option

This commit does not change behavior, it just changes code to handle -noconnect
values explicitly with IsArgNegated() instead of implicitly with IsArgSet(),
and adds comments to make it clear what behavior is intended when -noconnect is
specified.

ryanofsky committed:
doc: Add release notes summarizing negated option behavior changes.

ryanofsky committed:
Fix nonsensical -noseednode behavior

Treat specifying -noseednode the same as not specifying any -seednode value,
instead of enabling the seed node timeout and log messages, and waiting longer
to add other seeds.

ryanofsky committed:
Fix nonsensical bitcoin-cli -norpcwallet behavior

Treat specifying -norpcwallet the same as not specifying any -rpcwallet option,
instead of treating it like -rpcwallet=0 with 0 as the wallet name.

This restores previous behavior before 743077544b5420246ef29e0b708c90e3a8dfeeb6
from https://github.com/bitcoin/bitcoin/pull/18594, which inadvertently changed
it.

ryanofsky committed:
Fix nonsensical -norpcwhitelist behavior

Treat specifying -norpcwhitelist the same as not specifying -rpcwhitelist,
instead of behaving almost the same but flipping the default
-rpcwhitelistdefault value.

This is confusing because before this change if -norpcwhitelist was specified
it would block users from calling any RPC methods.

ryanofsky committed:
Fix nonsensical -notest behavior

Treat specifying -notest exactly the same as not specifying any
-test value, instead of complaining that it must be used with -regtest.

ryanofsky committed:
refactor: Avoid using IsArgSet() on -debug, -loglevel, and -vbparams list options

This commit does not change behavior, it just drops unnecessary IsArgSet()
calls for -debug, -loglevel, and -vbparams options. The calls are unnecessary
because GetArgs() already returns empty arrays if these arguments are not
specified.

ryanofsky committed:
Fix nonsensical -norpcbind and -norpcallowip behavior

Treat specifying -norpcbind and -norpcallowip the same as not specifying
-rpcbind or -rpcallowip, instead of failing to bind to localhost and failing to
show warnings.

Also add code comment to clarify what intent of existing code is.

ryanofsky committed:
Normalize inconsistent -noonlynet behavior

Treat specifying -noonlynet the same as not specifying -onlynet, instead of
marking all networks unreachable.

Before this change, specifying -noonlynet cleared list of reachable networks
and did not allow connecting to any network. It was basically an undocumented
synonym for -noconnect.

After this change, specifying -nononlynet just clears previously specifed
-onlynet options and allows connecting to all networks, restoring default
behavior as if no -onlynet options were specified.

Before this change, there was no way to restore default behavior once an
-onlynet option was specified. So for example, if a config file specifed
onlynet settings, they couldn't be reset on the command line without disabling
the entire config file.

The previous -noonlynet behavior wasn't neccessarily bad, but it was
undocumented, redundant with the -noconnect option, inconsistent with behavior
of other list options, and inconsistent with being able to use the command line
to selectively override config options. It was also probably unintended,
arising from use of the IsArgSet() method and its interaction with negated
options.

mzumsande reviewed (commented):
My understanding of the args parsing code is superficial, but why is changing `IsArgSet()` behavior not an option, instead of telling people not to use it anymore? Are there many spots that rely on it returning `true` in connection with a negated arg?

ryanofsky commented:
> > Make negating options the same as not setting them ... -noconnect
> 
> Maybe I am misunderstanding, but if you are trying to say that this treats `-noconnect` as if it never was set at all, then I tend to NACK

Nice catch! Definitely agree it would be a mistake to change `-noconnect` behavior, so I'll revert those changes and add comments to code to clarify the intent. Will also look into why `feature_config_args.py` test didn't catch this since it is testing `-noconnect` explicitly.

re: https://github.com/bitcoin/bitcoin/pull/30529#pullrequestreview-2201953036

> but why is changing `IsArgSet()` behavior not an option, instead of telling people not to use it anymore? Are there many spots that rely on it returning `true` in connection with a negated arg?

I don't think there is something wrong with `IsArgSet()` behavior. The function is doing what you'd expect it to do, but the it's just not a very useful function to call in most cases compared to alternatives. I think we could actually eliminate it pretty easily but did not want to expand the scope of this PR.

ryanofsky committed:
test: Add test to make sure -noconnect disables -dnsseed and -listen by default

Make sure -noconnect has same effect as -connect for disabling DNS seeding and
listening by default, and warning about -dnsseed being ignored with the -proxy
setting.

Initial implementation of https://github.com/bitcoin/bitcoin/pull/30529
accidentally broke this behavior, so having coverage may be useful to make sure
it does not break again.

Co-authored-by: Hodlinator <172445034+hodlinator@users.noreply.github.com>

ryanofsky commented:
Updated 3c17b800ca3273dcb5bd9b94c400396fdc80cd23 -> 02dde1ca402cc495949ac20f2b6d3c3626bd0144 ([`pr/listset.1`](https://github.com/ryanofsky/bitcoin/commits/pr/listset.1) -> [`pr/listset.2`](https://github.com/ryanofsky/bitcoin/commits/pr/listset.2), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/listset.1..pr/listset.2)) reverting -noconnect change and adding a test for the regression, also splitting up the main commit a bit and dropping the documentation change to make this PR more focused

stratospher reviewed (commented):
Concept ACK. nice to have more consistent behaviour for what to expect with the negated version of the config flags compared to on master.

ryanofsky commented on test/functional/feature_config_args.py:
```diff
@@ -319,6 +321,25 @@ def test_connect_with_seednode(self):
+            self.nodes[0].replace_in_config([("bind=", "#bind"), ("dnsseed=", "#dnsseed=")])
+            with self.nodes[0].assert_debug_log(expected_msgs=addcon_thread_started) as info:
+                self.restart_node(0, extra_args=[connect_arg])
+            self.nodes[0].replace_in_config([("#bind=", "bind"), ("#dnsseed=", "dnsseed=")])
```
re: https://github.com/bitcoin/bitcoin/pull/30529#discussion_r1719789856

> [ea7dbfb](https://github.com/bitcoin/bitcoin/commit/ea7dbfba8a62f3ba32b4386392721679d802e8ed):

Thanks! That might have caused problems if more tests were added after this

ryanofsky reviewed (commented):
Updated 10770c164485324af8acdd3b932b51d32862c41f -> a1c9d23ff587b84d37175b2993ea6760f9762177 ([`pr/listset.3`](https://github.com/ryanofsky/bitcoin/commits/pr/listset.3) -> [`pr/listset.4`](https://github.com/ryanofsky/bitcoin/commits/pr/listset.4), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/listset.3..pr/listset.4)) with suggested changes

naiyoma commented:
Tested ACK [https://github.com/bitcoin/bitcoin/pull/30529/commits/0e98e5b908dfc7cc8f6b4d2dca4d67b7fd8b5f7c](https://github.com/bitcoin/bitcoin/pull/30529/commits/0e98e5b908dfc7cc8f6b4d2dca4d67b7fd8b5f7c) 
Ran the functional test locally and test passes successfully, checks -noconnect  and -connect=0 disables `-listen` and` -dnsseed`.

naiyoma commented on test/functional/feature_config_args.py:
```diff
@@ -319,6 +321,25 @@ def test_connect_with_seednode(self):
+            for msg in dnsseed_ignored:
+                if msg not in info.log:
+                    raise AssertionError(f"Expected {msg!r} in -noconnect log message")
+
```
nit :  not sure about this but I think it would be useful to check that when -noconnect/connect=0 is set no outbound connections are made.

ryanofsky replied:
That would be nice and if I can figure out a way to check for that more directly I could drop the second half of this new test. It looks like there is a log message "%i block-relay-only anchors will be tried for connections" that gets printed when `m_use_addrman_outgoing` is false so maybe I could check for that when -noconnect is specified, and have better confirmation that noconnect is working correctly, than this `dnsseed_ignored` check provides.

I think the first half of the test checking for `dnsseed_disabled` and `listen_disabled` is useful in either case since it is providing test coverage for InitParameterInteraction code handling noconnect instead of AppInitMain code handling it.

achow101 commented:
cc @l0rinc @hodlinator

hodlinator reviewed (commented):
Concept ACK a1c9d23ff587b84d37175b2993ea6760f9762177

Good to fix weird negation corner cases.

(Very tempting to slap `DISALLOW_NEGATION` on non-list args for some reason, but I guess someone could be using `norpcwallet` to reset earlier values instead of `rpcwallet=`).

mzumsande reviewed (commented):
Concept ACK

Does this PR have the aspiration to fix all instances where it makes sense?

If so, I think that the pattern from  40ecf91068545ba55a6af5e2536a61c459334925 could also be applied to `-port` and `-rpcport` in `CheckHostPortOptions()`  or `-asmap`, which all fail init in master when negated and which could be used to negate earlier options if changed.

ryanofsky commented on src/init.cpp:
```diff
@@ -1918,17 +1919,25 @@ bool AppInitMain(NodeContext& node, interfaces::BlockAndHeaderTipInfo* tip_info)
+        // Warn that -dnsseed will be ignored if automatic connections are
+        // disabled.
         if (args.IsArgSet("-dnsseed") && args.GetBoolArg("-dnsseed", DEFAULT_DNSSEED) && args.IsArgSet("-proxy")) {
             LogPrintf("-dnsseed is ignored when -connect is used and -proxy is specified\n");
```
re: https://github.com/bitcoin/bitcoin/pull/30529#discussion_r1815689215

> Better to explain _why_ than _what_ (log message already explains _what_). To me it's not immediately obvious that we should avoid using DNS seeds when having a Tor SOCKS5 proxy?

Just dropped the new comments for now. There should be more chances to make improvements here later.

The comment was saying why the dnsseed is ignored: because automatic connections are disabled. The log message says the seed is ignored because -connect is specified, but it is not obvious (IMO) that specifying -connect disables other connections. 

The -proxy part of this check seems buggy to me and I mentioned that in the commit message, but didn't look into the history, and in any case don't want to change the behavior in this PR.

hodlinator replied:
Unintended negation in commit message of 0a954ce80cf9d9744eaf213e3b81d4016b19bf88?
"Not sure if it really makes sense >not< to warn about seednode being ignored if
-noconnect is specified, and only to warn about -dnsseed being ignored when
-proxy is specified, but these behaviors are not changed from before."

ryanofsky commented on src/torcontrol.cpp:
```diff
@@ -403,7 +403,7 @@ void TorController::get_socks_cb(TorControlConnection& _conn, const TorControlRe
 
     const bool onion_allowed_by_onlynet{
-        !gArgs.IsArgSet("-onlynet") ||
+        gArgs.GetArgs("-onlynet").empty() ||
```
re: https://github.com/bitcoin/bitcoin/pull/30529#discussion_r1815702002

> Use local variable.

Good catch, updated

ryanofsky reviewed (commented):
Thanks for the reviews!

Rebased a1c9d23ff587b84d37175b2993ea6760f9762177 -> 54ad580775b8379a62b05b1f2cdf8b7fde993306 ([`pr/listset.4`](https://github.com/ryanofsky/bitcoin/commits/pr/listset.4) -> [`pr/listset.5`](https://github.com/ryanofsky/bitcoin/commits/pr/listset.5), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/listset.4-rebase..pr/listset.5)) due to conflict with #31212 and to make suggested changes.

---


re: https://github.com/bitcoin/bitcoin/pull/30529#pullrequestreview-2400135259

> Does this PR have the aspiration to fix all instances where it makes sense?
> 
> If so, I think that the pattern from [40ecf91](https://github.com/bitcoin/bitcoin/commit/40ecf91068545ba55a6af5e2536a61c459334925) could also be applied to `-port` and `-rpcport` in `CheckHostPortOptions()` or `-asmap`, which all fail init in master when negated and which could be used to negate earlier options if changed.

Need to look into these. There are definitely more places where behavior could be improved. This PR was specifically trying to cover all the cases where `IsArgSet` and `GetArgs` were being used together with goal of simplifying #17783 and doesn't look at most cases beyond that.

hodlinator commented on src/init.cpp:
```diff
@@ -1918,17 +1919,25 @@ bool AppInitMain(NodeContext& node, interfaces::BlockAndHeaderTipInfo* tip_info)
+        // Assign addresses of trusted nodes, if any.
         if (connect.size() != 1 || connect[0] != "0") {
             connOptions.m_specified_outgoing = connect;
         }
```
Yeah, I think my understanding of how list-args are treated is less solid than yours. Calling `connect`-values "trusted" added some context, but is kind of implied. Removing the comment was a better choice than taking my suggestion. :+1:

hodlinator reviewed (commented):
Code reviewed 54ad580775b8379a62b05b1f2cdf8b7fde993306

### Commit message 594acdb8e23509350a5d3a9db18b76823ca5bce3

> - Treats specifying -noonlynet exactly the same as not specifying -onlynet
>  instead of marking all networks unreachable.

It strikes me that maybe someone found the old behavior of disabling all networks though negating this arg useful. One cannot disable all networks by just passing `-onlynet` without a value since it results in:
```
Error: Unknown network specified in -onlynet: ''
```
But maybe `-noonlynet` is counter-intuitive enough for disabling all networks that your change is justified all the same.

hodlinator commented on test/functional/test_framework/test_node.py:
```diff
@@ -498,7 +499,8 @@ def assert_debug_log(self, expected_msgs, unexpected_msgs=None, timeout=2):
         prev_size = self.debug_log_size(encoding="utf-8")  # Must use same encoding that is used to read() below
 
-        yield
+        info = types.SimpleNamespace()
+        yield info
```
(Not applicable if https://github.com/bitcoin/bitcoin/pull/30529#discussion_r1886540011 is followed).

ryanofsky committed:
Fix nonsensical -noasmap behavior

Instead of failing with "fread failed: iostream error" error when -noasmap is
specified, just don't load an asmap file.

ryanofsky reviewed (commented):
Thanks for the new review!

Updated 54ad580775b8379a62b05b1f2cdf8b7fde993306 -> 80608ba5d282ae8713ea0136ea9a0208b254c082 ([`pr/listset.5`](https://github.com/ryanofsky/bitcoin/commits/pr/listset.5) -> [`pr/listset.6`](https://github.com/ryanofsky/bitcoin/commits/pr/listset.6), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/listset.5..pr/listset.6)) improving various comments and adding a new commit to handle `-noasmap` based on Martin's suggestion https://github.com/bitcoin/bitcoin/pull/30529#pullrequestreview-2400135259.

---

re: https://github.com/bitcoin/bitcoin/pull/30529#pullrequestreview-2400135259

> Does this PR have the aspiration to fix all instances where it makes sense?
>
> If so, I think that the pattern from [40ecf91](https://github.com/bitcoin/bitcoin/commit/40ecf91068545ba55a6af5e2536a61c459334925) could also be applied to `-port` and `-rpcport` in `CheckHostPortOptions()` or `-asmap`, which all fail init in master when negated and which could be used to negate earlier options if changed.

Looked into each of these (`-noport` `-norpcport` and `-asmap`) and I think all could be improved but the port options would be difficult to fix in a clean way without #31260 because of their interactions with bind options. I added a new commit to this PR to fix `-noasmap` behavior, and listed other ideas for improving these options below.

- Currently `-noasmap` fails with error "fread failed: iostream error" because it tries to open an empty file path. I added a new commit to fix this and make `-noasmap` just not load a file. (Separately I think there are other problems with the `-asmap` option implemention. The `-asmap` option is documented have a default value of `ip_asn.map` but this is not accurate, because no asmap file is actually loaded by default, and if an `ip_asn.map` file happens to exist it is just ignored. In order to cause an ASN map to be loaded you have to explicitly pass `-asmap`. I don't think it makes sense to have a default asmap path if it will not actually be loaded by default. Better behavior would be to treat `ip_asn.map` like `bitcoin.conf` and just load it by default as long as it exists and another path is not specified.)

- Currently `-norpcport` is treated like `-rpcport=0` and always results in "Invalid port specified in -rpcport: '0'" errors (whether or not it is specified in combination with `-rpcbind`). I think more ideally, `-norpcport` would not be an error, it would just be accept and stop the RPC server binding to a port. Additionally, I don't think current behavior of `-norpcbind` causing the RPC server to bind to localhost makes sense. Based on behavior of `-noconnect` and the meaning of the word "no" I would expect `-norpcbind` to cause RPC server not to bind to an address. Better behavior would be easier to implement in a clean way with #31260 so settings could be given explicit types. It would make sense for the `-rpcport` setting to have type `std::optional<uint16_t>` and for the `-rpcbind` setting to have type `std::optional<std::vector<std::string>>`, and then the following code could be implemented
  ```c++
  // -rpcport is a more concise but less flexible alternative to -rpcbind, and
  // -rpcbind takes precedence if the options conflict. Negation is allowed, with
  // -norpcbind or -norpcport causing the RPC server not to bind to an address.
  const auto rpc_port = settings::RpcPort::Get(args);
  const auto rpc_bind = settings::RpcBind::Get(args);
  const uint16_t default_rpc_port = rpc_port && *rpc_port ? *rpc_port : base_params.RPCPort();
  std::vector<std::pair<std::string, uint16_t>> endpoints;
  if (rpc_bind) {
     for(const auto std::string& addr : *rpc_bind) {
        endpoints.emplace_back(SplitHostPort(addr, default_rpc_port));
     }
  } else if (rpc_port != 0) {
      endpoints.emplace_back("::1", default_rpc_port);
      endpoints.emplace_back("127.0.0.1", default_rpc_port);
  }
  ```

- Currently `-noport` is just treated as `-port=0` and always results in "Invalid port specified in -port: '0'". Also currently `-nobind` binds to "0.0.0.0" instead of not binding to an address. IMO it would make more sense for `-noport` or `-nobind` to disable binding, in the same way suggested for `-rpcport` and `-rpcbind above.

---

re: https://github.com/bitcoin/bitcoin/pull/30529#pullrequestreview-2505673099

> It strikes me that maybe someone found the old behavior of disabling all networks though negating this arg useful.

I think if someone wanted to not connect to any network they would use `-noconnect` instead of `-noonlynet`. I'm not sure how someone would actually discover that `-noonlynet` disables network connections, since I'd think a more logical expectation would be that `-onlynet` disables networks so `-noonlynet` re-enables them.

hodlinator commented on doc/release-notes-30529.md:
```diff
@@ -0,0 +1,4 @@
+Configuration
+-------------
+
+Some corner cases handling negated list options `-norpcwhitelist`, `-norpcallowip`, `-norpcbind`, `-nobind`, `-nowhitebind`, `-noexternalip`, `-noonlynet`, `-noseednode`, `-nosignetchallenge`, `-nosignetseednode`, `-notest`, `-norpcwallet`, and `-no have been fixed. Now negating these options is the same as not specifying them.
```
Typo:
```
`-norpcwallet`, and `-no have
```

Also not sure this should be included in this sentence assuming it is `-asmap` since it was not used as a list arg?

hodlinator reviewed (commented):
Code reviewed 80608ba5d282ae8713ea0136ea9a0208b254c082

`git range-diff master 54ad580 80608ba`

danielabrozzoni commented on doc/release-notes-30529.md:
```diff
@@ -0,0 +1,4 @@
+Configuration
+-------------
+
+Some corner cases handling negated list options `-norpcwhitelist`, `-norpcallowip`, `-norpcbind`, `-nobind`, `-nowhitebind`, `-noexternalip`, `-noonlynet`, `-noseednode`, `-nosignetchallenge`, `-nosignetseednode`, and `-notest` have been fixed. Now negating these options is the same as not specifying them.
```
In b8c83c3a90452a35f63daac70d7cdf3ad0f2e280:

The commit message and release notes say that there are corner cases being fixed in `-nosignetseednode` and `-nosignetchallenge`, but I can't seem to reproduce them, as the code disallows the two options:

```
❯ ./build/src/bitcoind -nosignetchallenge
Error: Error parsing command line arguments: Negating of -signetchallenge is meaningless and therefore forbidden
                                                                                                                                            
❯ ./build/src/bitcoind -nosignetseednode 
Error: Error parsing command line arguments: Negating of -signetseednode is meaningless and therefore forbidden
```

Am I doing something wrong?
If that's not the case, can you specify in the commit message that for those two calls the behavior is not impacted, but the code is refactored for clarity?

danielabrozzoni reviewed (commented):
Concept ACK - I still haven't reviewed every commit, but this seems a nice cleanup overall.

ryanofsky committed:
doc: Add some general documentation about negated options

ryanofsky reviewed (commented):
Note to previous reviewers: I broke up and rearranged all the commits here and rewrote most of the descriptions to make the purpose of the PR clearer, but there are no significant code changes since the last push (just a small refactoring in the `-onlynet` code and a suggested test simplification).

---

Updated 80608ba5d282ae8713ea0136ea9a0208b254c082 -> d4b02f04c8960f9d1fa9ff69977b21eccb104d41 ([`pr/listset.6`](https://github.com/ryanofsky/bitcoin/commits/pr/listset.6) -> [`pr/listset.7`](https://github.com/ryanofsky/bitcoin/commits/pr/listset.7), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/listset.6..pr/listset.7)) rearranging commits and improving documentation
Updated d4b02f04c8960f9d1fa9ff69977b21eccb104d41 -> a0941aaa0336440458ee4343118977d3989943eb ([`pr/listset.7`](https://github.com/ryanofsky/bitcoin/commits/pr/listset.7) -> [`pr/listset.8`](https://github.com/ryanofsky/bitcoin/commits/pr/listset.8), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/listset.7..pr/listset.8)) fixing typos and making  a suggested test simplification.
Updated a0941aaa0336440458ee4343118977d3989943eb -> eb325bc0efd3f999189e155ba836be92e6cc9af7 ([`pr/listset.8`](https://github.com/ryanofsky/bitcoin/commits/pr/listset.8) -> [`pr/listset.9`](https://github.com/ryanofsky/bitcoin/commits/pr/listset.9), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/listset.8..pr/listset.9)) simplifying bitcoin-cli code and adding a test and making other small cleanups

hodlinator reviewed (approved):
ACK eb325bc0efd3f999189e155ba836be92e6cc9af7

Nice break up of commits for correcting each `-arg`.

In my case I still reviewed mostly through:
``` 
₿ git diff 2eccb8bc5e2245e42e05b8df3d9aa264b5f21a9a 80608ba5d282ae8713ea0136ea9a0208b254c082 > old
₿ git diff 2eccb8bc5e2245e42e05b8df3d9aa264b5f21a9a eb325bc0efd3f999189e155ba836be92e6cc9af7 > new
₿ meld old new &
```

*feature_config_args.py* - Nice that we are no longer relying on "leaked" `connect_arg` from prior `for`-loop, instead modifying the loop body now. Thanks for taking my feedback into account.

Good additions to `bitcoin-conf.md`.

Good to see added `-norpcwallet` test and `RpcWalletName()`-function.

Passed unit & functional tests locally.

danielabrozzoni reviewed (approved):
tACK eb325bc0efd3f999189e155ba836be92e6cc9af7 - I reviewed the code and manually verified the behavior of the affected RPC arguments, both old and new.

Thank you for splitting the changes into several separate commits; it made the review process much easier! :)

danielabrozzoni commented on src/init/common.cpp:
```diff
@@ -73,14 +72,12 @@ util::Result<void> SetLoggingLevel(const ArgsManager& args)
+        // Special-case: if -debug=0/-debug=none is set, turn off debugging messages
         const std::vector<std::string> categories = args.GetArgs("-debug");
 
         if (std::none_of(categories.begin(), categories.end(),
```
Would one of you like to take this on, or can I give it a shot? :)

The way I understand this, we would show a warning if the user sets both debug=0/none and some other debug categories, but we wouldn’t warn if -nodebug is set with other debug categories. This way, -nodebug and -debug=something can still be used together in the CLI to reset previous debugging options without triggering a warning. Does that make sense?

ryanofsky replied:
re: https://github.com/bitcoin/bitcoin/pull/30529/commits/76fc51774e984c81512f2dd40dbaf42d2bbda18d#r1930837849

Definitely feel free to improve this. I think the most elegant solution instead of treating -debug=0 and -debug=none differently from -nodebug would be to treat them all the same and whenever any of these are specified just ignore any previous options.

Ability to reset the list and append new options afterward without having to choose one or the other can be useful when you have a wrapper script that passes some default options and you want to be able to pass more options to override them. FWIW, when I gave this suggestion to chatgpt it suggested:

```c++
auto last_negated = std::ranges::find_if(categories | std::views::reverse, 
    [](const std::string& cat) { return cat == "0" || cat == "none"; });

auto range_to_process = (last_negated == categories.rend()) 
    ? categories 
    : std::ranges::subrange(last_negated.base(), categories.end());

for (const auto& cat : range_to_process) {
    if (!LogInstance().EnableCategory(cat)) {
        return util::Error{strprintf(_("Unsupported logging category %s=%s."), "-debug", cat)};
    }
}
```

hodlinator replied:
Happy to review. ryanofsky's suggestion feels more consistent with behavior of scalar/non-list settings, so I think it's worth trying out before my original suggestion.

danielabrozzoni replied:
Thank you to both of you! I opened https://github.com/bitcoin/bitcoin/pull/31767 :)

> FWIW, when I gave this suggestion to chatgpt it suggested:

I ended up using a for loop instead of ranges, but I kept the rest; I put you as the commit co-author, let me know if you'd rather not be a co-author though!

hodlinator commented on test/functional/feature_config_args.py:
```diff
@@ -391,6 +393,25 @@ def test_connect_with_seednode(self):
+            self.restart_node(0, extra_args=[connect_arg])
+        for msg in [dnsseed_disabled, listen_disabled]:
+            if msg not in info.log:
+                raise AssertionError(f"Expected {msg!r} in -noconnect log message")
```
I agree when it comes to `assert()`s in C/C++, guess I haven't fully transferred that to Python yet, but this discussion helped. What do you think of doing a scripted diff renaming our `assert_*()`-functions and their usage to `check_*()` to signal that they won't be optimized out?

hodlinator replied:
GoogleTest which seems like a slightly more desirable framework than Boost Unit Tests also uses [`ASSERT_*`-style](https://google.github.io/googletest/reference/assertions.html) statements I realize, so maybe keeping with that term is good in the long run.

What we're missing may be generic `assert_true(x)` & others that can be used where the special Python `assert x` feature is a bad fit.

hodlinator replied:
Was about to add built-in `assert` in a small fix in #31874, took the opportunity to tentatively add `assert_true()`.

ryanofsky commented:
Rebased eb325bc0efd3f999189e155ba836be92e6cc9af7 -> a85e8c0e6158fad2408bda5cb1e36da707eb081b ([`pr/listset.9`](https://github.com/ryanofsky/bitcoin/commits/pr/listset.9) -> [`pr/listset.10`](https://github.com/ryanofsky/bitcoin/commits/pr/listset.10), [compare](https://github.com/ryanofsky/bitcoin/compare/pr/listset.9-rebase..pr/listset.10)) due to conflict with #31767

danielabrozzoni reviewed (commented):
re-ACK a85e8c0e6158fad2408bda5cb1e36da707eb081b

hodlinator reviewed (approved):
re-ACK a85e8c0e6158fad2408bda5cb1e36da707eb081b

Resolved trivial conflict with #31767 since [my other ACK](https://github.com/bitcoin/bitcoin/pull/30529#pullrequestreview-2546259926).

achow101 commented:
ACK a85e8c0e6158fad2408bda5cb1e36da707eb081b
